modules/node-shared/src/main/scala/org/tessellation/node/shared/domain/queue/ViewableQueue.scala

```diff
@@ -0,0 +1,49 @@
+package org.tessellation.node.shared.domain.queue
+
+import cats.effect.std.Queue
+import cats.effect.{Concurrent, Ref}
+import cats.implicits._
+
+trait ViewableQueue[F[_], A] {
+  def offer(item: A): F[Unit]
+
+  def take: F[A]
+
+  def tryOfferN(items: List[A]): F[List[A]]
+
+  def tryTakeN(n: Option[Int]): F[List[A]]
+
+  def view: F[List[A]]
+}
+
+object ViewableQueue {
+
+  def unbounded[F[_]: Concurrent, A]: F[ViewableQueue[F, A]] =
+    Queue.unbounded[F, A].flatMap(q => make(q))
+
+  def make[F[_]: Concurrent, A](queue: Queue[F, A]): F[ViewableQueue[F, A]] =
+    Ref.of[F, List[A]](List.empty).map { state =>
```

Comment:
I realized there may be an issue with synchronization between the `Ref` and the `Queue`. I'll investigate other implementations that wrap these two effects into one atomic update.

Reply:
added a Mutex lock and wrapped write operations to ensure atomicity